src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs

```diff
@@ -1,348 +1,370 @@
+                throw new InvalidOperationException($"Can't change existing formula: {name}");
+            }
+
+            var check = Check(expr._expression, expr._schema);
+            check.ThrowOnErrors();
+            var binding = check._binding;
+
+            // We can't have cycles because:
+            // - formulas can only refer to already-defined values
+            // - formulas can't be redefined.  
+            HashSet<string> dependsOn = check.TopLevelIdentifiers;
+
+            var type = FormulaType.Build(binding.ResultType);
+            var info = new RecalcFormulaInfo
+            {
+                _dependsOn = dependsOn,
+                _type = type,
+                _binding = binding,
+                _onUpdate = onUpdate
+            };
+
+            Formulas[name] = info;
+
+            foreach (var x in dependsOn)
+            {
+                Formulas[x]._usedBy.Add(name);
+            }
+
+            Recalc(name);
+        }
+
+        /// <summary>
+        /// Get intellisense from the formula.
+        /// </summary>
+        public IIntellisenseResult Suggest(string expression, FormulaType parameterType, int cursorPosition)
+        {
+            var result = this.CheckInternal(expression, parameterType, intellisense: true);
+            var binding = result._binding;
+            var formula = result._formula;
+
+            var context = new IntellisenseContext(expression, cursorPosition);
+            var intellisense = IntellisenseProvider.GetIntellisense(_powerFxConfig.EnumStore);
+            var suggestions = intellisense.Suggest(context, binding, formula);
+
+            return suggestions;
+        }
+
+        // Trigger a recalc on name and anything that depends on it. 
+        // Invoke on Update callbacks. 
+        private void Recalc(string name)
+        {
+            var r = new RecalcEngineWorker(this);
+            r.Recalc(name);
+        }
+
+        public void DeleteFormula(string name)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Get the current value of a formula. 
+        /// </summary>
+        /// <param name="name"></param>
+        /// <returns></returns>
+        public FormulaValue GetValue(string name)
+        {
+            var fi = Formulas[name];
+            return fi._value;
+        }
+
+        /// <summary>
+        /// Returns the parse tree (AST) that Power Fx builds for this expression. 
+        /// </summary>
+        /// <param name="expressionText">The expression we plan to parse</param>
+        /// <param name="recordType">All types available at execution time</param>
+        /// <param name="expression">The parsed expression. Expression will be null if !CheckResult.IsSuccess</param>
+        /// <returns>CheckResult</returns> Returns any errors detected  
+
+        public CheckResult ParseExpression(string expressionText, RecordType recordType, out IExpression expression)

```

Comment:
> out IExpression expression [](http://example.com/codeflow?start=88&length=27)

Should this just be another propery on CheckResult, and we use the existing Check method?